torch/csrc/jit/mobile/versioned_operators.cpp

```diff
@@ -0,0 +1,185 @@
+#include <c10/util/Optional.h>
+#include <torch/csrc/jit/mobile/versioned_operators.h>
+
+#include <unordered_map>
+
+// This file is to handle operator versioning in mobile.
+// How does it work?
+// At top level, with a tuple of (operator_name, op_version_in_model,
+// bytecode_version_in_model), check the compatibility for that operator. If
+// it's not compatible, throw an error with reason. If it's compatible, do
+// compatibility treatments if necessary and return the corresponding operator
+// functor that is compatible to this runtime. The logic is implemented in
+// function operator_resolver().
+//
+// Under the hood, a static table,
+// op_version_table, is saved for current runtime. The keys are the operator
+// names and values are a container with version numbers and their metadata. The
+// incoming operator with version is searched in the table to find the match.
+// Note that the size of op_version_table is small because it only contains
+// operators that have more than one versions. For most of the operators, they
+// fall through to the existing implementation with default version 0.
+
+namespace torch {
+namespace jit {
+namespace mobile {
+namespace {
+
+// The version info is limited to combinations of two situations:
+//   * Route to another op name
+//   * Push additional default values
+// It’s fine to update the content of version info, to handle more situations in
+// fugure. Since it’s implementation details, it should not affect
+// compatibility.
+struct VersionInfo {
+  VersionInfo(
+      int64_t version,
+      const c10::OperatorName& name,
+      c10::optional<IValue> default_val)
+      : version(version), name(name), default_val(default_val) {}
+  int64_t version;
+  c10::OperatorName name;
+  c10::optional<IValue> default_val;
+};
+
+struct VersionInfoContainer {
+  VersionInfoContainer(const std::vector<VersionInfo>& version_vec)
+      : version_vec(version_vec) {}
+  int64_t min_version() const {
+    return version_vec.begin()->version;
+  }
+
+  int64_t max_version() const {
+    return version_vec.rbegin()->version + 1;
```

Comment:
Wny is there a +1 here?

Reply:
The table saves the older versions and their compatibility actions. For example, if an op has different treatments for version from 0 to 2, this runtime could also handle 3, which is the default version. For default version, it's like all other operators not in the table. There's no special handling so no need to save here.